File: app/src/main/java/com/alphabyte/maths/helper/PreferenceHelper.java
package com.alphabyte.maths.helper;

import android.content.Context;
import android.util.Log;

import com.alphabyte.maths.models.Favourite;
import com.alphabyte.maths.models.Maths;
import com.alphabyte.maths.models.TopicList;
import com.google.gson.Gson;
import com.orhanobut.hawk.Hawk;

import java.util.ArrayList;
import java.util.List;

public class PreferenceHelper {
    Context mContext;
    private String FAVOURITE_TOPIC = "favourite";
    private String DARK_THEME = "darkTheme";
    private String FONT_SIZE = "fontSize";
    Gson gson;
    boolean darkTheme;
    int fontSize;


    String favourite;


    public PreferenceHelper(Context mContext) {
        this.mContext = mContext;
        Hawk.init(mContext).build();
        gson = new Gson();
        if(Hawk.count() != 0){
            favourite = Hawk.get(FAVOURITE_TOPIC,"");
        }else{
            favourite = "";
        }
        darkTheme = false;
        fontSize = 12;
    }

    public boolean isDarkTheme() {
        darkTheme = Hawk.get(DARK_THEME,false);
        return darkTheme;
    }

    public void setDarkTheme(boolean darkTheme) {
        Hawk.put(DARK_THEME,darkTheme);
        this.darkTheme = darkTheme;
    }

    public int getFontSize() {
        fontSize = Hawk.get(FONT_SIZE,12);
        return fontSize;
    }

    public void setFontSize(int fontSize) {
        Hawk.put(FONT_SIZE,fontSize);
        this.fontSize = fontSize;
    }

    public String getFavouriteList() {
        favourite = Hawk.get(FAVOURITE_TOPIC,"");
        return favourite;
    }

    public void addTopic(Maths.Topic topic,String fileName) {
        List<TopicList.TopicDetails> topicList;
        TopicList.TopicDetails topicDetails = new TopicList.TopicDetails();
        topicDetails.setTopic_name(topic.getTopic_name());
        topicDetails.setTopic_file_name(fileName);
        int flag = 0;
        if(favourite != null && favourite.length() != 0){
            Favourite fav = gson.fromJson(favourite, Favourite.class);
            topicList = fav.getTopicList();
            for(TopicList.TopicDetails t : topicList){
                if(t.getTopic_name().equals(topicDetails.getTopic_name())){
                    flag = 1;
                }
            }
            if(flag == 0){
                topicList.add(topicDetails);
            }
            fav.setTopicList(topicList);
            favourite = gson.toJson(fav);
            Hawk.put(FAVOURITE_TOPIC,favourite);
        }else {
            Favourite fav = new Favourite();
            topicList = new ArrayList<>();
            topicList.add(topicDetails);
            fav.setTopicList(topicList);
            favourite = gson.toJson(fav);
            Hawk.put(FAVOURITE_TOPIC,favourite);
        }
    }

    public void removeTopic(Maths.Topic topic,String filename){
        List<TopicList.TopicDetails> topicList;
        List<TopicList.TopicDetails> newList;
        newList = new ArrayList<>();
        String output;
        int index = -1;
        String favString = getFavouriteList();
        Favourite fav = gson.fromJson(favString,Favourite.class);
        topicList = fav.getTopicList();
        for(TopicList.TopicDetails t: topicList) {
            if (!t.getTopic_name().equals(topic.getTopic_name())) {
                newList.add(t);
            }
        }

        fav.setTopicList(newList);
        output = gson.toJson(fav);
        Log.i("removeTopic",topic.getTopic_name() + "  " + output );
        favourite = output;
        Hawk.put(FAVOURITE_TOPIC,favourite);
    }


}

File: app/src/main/java/com/alphabyte/mathsformulabook/helper/PreferenceHelper.kt
package com.alphabyte.mathsformulabook.helper

import android.content.Context
import android.util.Log
import com.alphabyte.mathsformulabook.models.Favourite
import com.alphabyte.mathsformulabook.models.Maths
import com.alphabyte.mathsformulabook.models.TopicList
import com.google.gson.Gson
import com.orhanobut.hawk.Hawk
import java.util.*

class PreferenceHelper(internal var mContext: Context) {
    private val FAVOURITE_TOPIC = "favourite"
    private val DARK_THEME = "darkTheme"
    private val FONT_SIZE = "fontSize"
    private var gson: Gson
    private var darkTheme: Boolean = false
    private var fontSize: Int = 0


    private var favourite: String? = null

    var isDarkTheme: Boolean
        get() {
            darkTheme = Hawk.get(DARK_THEME, false)
            return darkTheme
        }
        set(darkTheme) {
            Hawk.put(DARK_THEME, darkTheme)
            this.darkTheme = darkTheme
        }

    val favouriteList: String?
        get() {
            favourite = Hawk.get(FAVOURITE_TOPIC, "")
            return favourite
        }


    init {
        Hawk.init(mContext).build()
        gson = Gson()
        if (Hawk.count() != 0L) {
            favourite = Hawk.get(FAVOURITE_TOPIC, "")
        } else {
            favourite = ""
        }
        darkTheme = false
        fontSize = 12
    }

    fun getFontSize(): Int {
        fontSize = Hawk.get(FONT_SIZE, 12)
        return fontSize
    }

    fun setFontSize(fontSize: Int) {
        Hawk.put(FONT_SIZE, fontSize)
        this.fontSize = fontSize
    }

    fun addTopic(topic: Maths.Topic, fileName: String) {
        val topicList: MutableList<TopicList.TopicDetails>
        val topicDetails = TopicList.TopicDetails()
        topicDetails.topic_name = topic.topic_name
        topicDetails.topic_file_name = fileName
        var flag = 0
        if (favourite != null && favourite!!.length != 0) {
            val fav = gson.fromJson(favourite, Favourite::class.java)
            topicList = fav.topicList
            for (t in topicList) {
                if (t.topic_name == topicDetails.topic_name) {
                    flag = 1
                }
            }
            if (flag == 0) {
                topicList.add(topicDetails)
            }
            fav.topicList = topicList
            favourite = gson.toJson(fav)
            Hawk.put<String>(FAVOURITE_TOPIC, favourite)
        } else {
            val fav = Favourite()
            topicList = ArrayList<TopicList.TopicDetails>()
            topicList.add(topicDetails)
            fav.topicList = topicList
            favourite = gson.toJson(fav)
            Hawk.put<String>(FAVOURITE_TOPIC, favourite)
        }
    }

    fun removeTopic(topic: Maths.Topic, filename: String) {
        val topicList: List<TopicList.TopicDetails>
        val newList: MutableList<TopicList.TopicDetails>
        newList = ArrayList<TopicList.TopicDetails>()
        val output: String
        val index = -1
        val favString = favouriteList
        val fav = gson.fromJson(favString, Favourite::class.java)
        topicList = fav.topicList
        for (t in topicList) {
            if (t.topic_name != topic.topic_name) {
                newList.add(t)
            }
        }

        fav.topicList = newList
        output = gson.toJson(fav)
        Log.i("removeTopic", topic.topic_name + "  " + output)
        favourite = output
        Hawk.put<String>(FAVOURITE_TOPIC, favourite)
    }


}
